src/gordon_gcp/plugins/enricher.py

```diff
@@ -49,11 +56,87 @@ class GCEEnricher:
         self.config = config
         self.success_channel = success_channel
         self.error_channel = error_channel
+        self._http_client = http_client
+
+    async def _poll_for_ext_ip(self, resource_name):
+        exception = None
+        backoff = 2
+        base_url = f'https://compute.googleapis.com/compute/v1/{resource_name}'
+
+        for attempt in range(1, self.config['retries'] + 1):
+            try:
+                logging.debug(f'Attempt {attempt}: fetching {base_url}')
+                instance_data = await self._http_client.get_json(base_url)
+                return instance_data['networkInterfaces'][0][
+                    'accessConfigs'][0]['natIP']
+            except exceptions.GCPHTTPError as e:
+                exception = e
+                break
+            except (KeyError, IndexError) as e:
+                exception = e
+                if attempt == self.config['retries']:
+                    continue
+                await asyncio.sleep(backoff)
+                backoff = backoff ** 2
+
+        msg = f'Could not get external IP for {resource_name}: {exception}'
+        raise exceptions.GCPGordonError(msg)
+
+    async def _create_A_rrecord(self, data):
+        external_ip = await self._poll_for_ext_ip(data['resourceName'])
+        dns_zone = self.config['dns_zone']
+        hostname = data['resourceName'].split('/')[-1]
+
+        name = '.'.join([hostname, dns_zone])
+        logging.debug(f'Creating A record: {name} -> {external_ip}')
+        return {
+            'name': name,
+            'rrdatas': [external_ip],
+            'type': 'A'
+        }
+
+    async def process(self, event_message):
+        """
+        Enrich message with extra information and send it to publisher.
+
+        When a message is successfully processed, it is passed to the
+        :obj:`self.success_channel`. However, if there is a problem
+        during processing, the message is passed to the
+        :obj:`self.error_channel`.
+
+        Args:
+            event_message (.GEventMessage): message requiring
+                additional information.
+        """
+        data = event_message.data
+        try:
+            a_record = await self._create_A_rrecord(data)
+        except exceptions.GCPGordonError as e:
+            msg = ('Encountered error while enriching message, sending message'
+                   f' to error channel: {e}')
+            logging.warning(msg)
+            await self.error_channel.put(event_message)
+            return
+        logging.debug(f'Created resource record {a_record}')
+        event_message.data['resourceRecords'].append(a_record)
+        msg = f'Enriched msg {event_message.msg_id} with resourceRecords'
+        event_message.append_to_history(msg, self.phase)
+        await self.update_phase(event_message, 'publish')
+        await self.success_channel.put(event_message)
+
+    async def update_phase(self, event_message, phase):
```

Comment:
> Maybe this should live on the message itself? since it's acting on the message?

Talked about about this IRL: this will be moved to the msg when working on core routing logic.